pkg/apis/pipeline/v1alpha1/task_validation.go

```diff
@@ -132,67 +132,110 @@ func validateInputParameterVariables(steps []corev1.Container, inputs *Inputs) *
 			parameterNames[p.Name] = struct{}{}
 		}
 	}
-	return validateVariables(steps, "params", parameterNames)
+	return validateVariables(steps, "params", "inputs.", parameterNames)
 }
 
 func validateResourceVariables(steps []corev1.Container, inputs *Inputs, outputs *Outputs) *apis.FieldError {
-	resourceNames := map[string]struct{}{}
+	inputVars, err := getInputResourceVariables(inputs)
+	if err != nil {
+		return err
+	}
+	outputVars, err := getOutputResourceVariables(outputs)
+	if err != nil {
+		return err
+	}
+	err = validateVariables(steps, "resources", "inputs.", inputVars)
+	if err != nil {
+		return err
+	}
+	return validateVariables(steps, "resources", "outputs.", outputVars)
+}
+
+func getInputResourceVariables(inputs *Inputs) (map[string]struct{}, *apis.FieldError) {
+	vars := map[string]struct{}{}
 	if inputs != nil {
 		for _, r := range inputs.Resources {
-			resourceNames[r.Name] = struct{}{}
+			attrs, err := AttributesFromType(r.Type)
+			if err != nil {
+				return nil, &apis.FieldError{
+					Message: fmt.Sprintf("invalid resource type %s", r.Type),
+					Paths:   []string{"taskspec.inputs.resources." + r.Name},
+					Details: err.Error(),
+				}
+			}
+			for _, a := range attrs {
+				rv := r.Name + "." + a
+				vars[rv] = struct{}{}
+			}
 		}
 	}
+	return vars, nil
+}
+
+func getOutputResourceVariables(outputs *Outputs) (map[string]struct{}, *apis.FieldError) {
```

Comment:
i think it's worth having unit tests for `getOutputResourceVariables` and `getInputResourceVariables` specifically - that might also help address @sbwsg 's feedback about knowing exactly what changed, b/c we'd be able to see it in the test cases

Reply:
This changed slightly and has been reduced to one function, `getResourceVariables`. Your comment is still completely valid though. I have an idea for a test that could be constructive for `getResourceVariables`, but I want to put a bit more thought into it so that it's not tightly coupled with the current structure of a given resource type. Will circle back to this tomorrow.